### DISCUSSION TITLE
Enhance Content Collection: Strict Content Heading

### DISCUSSION THREAD

**Discussion by naranyala** [withastro/roadmap#722] [Proposal]
### Body

# Enhance Content Collection: Strict Content Heading

First of all thank you Astro Team and Community for making the web better.

Inspired by errors that occur when content frontmatter doesn't match the schema, this type of error is shown to help identify content heading structure. Such errors typically arise when the content lacks the proper h1, h2, h3 structure. For instance, if the content (markdown) only contains h2, h2, and h2 headings without an h1 at the top, or in cases where nested headings lack clear parent headings.

## Demo Errors

The error will occur in cases such as:
- When an h2 heading doesn't have an h1 parent or in similar nested scenarios.
- After an h1 heading, there is no h2, but an h3 is found.

![example-error-01](https://github.com/withastro/roadmap/assets/25216912/e5c608ca-7a84-4b5a-b481-aa1e96dcbb08)

![example-error-02](https://github.com/withastro/roadmap/assets/25216912/07a318b6-7b2f-48fd-914f-ba3eacbf0ca4)

## Ideas

In case we want to create a "table-of-contents" and be able to search for heading content, similar to the GitHub README feature. The primary difference compared to the current result from the content-collection is the new "headings" array for each markdown content (check stackblitz demo). It would be convenient if this functionality were integrated as a built-in feature in Astro's Content Collection.

## Motivation

My primary goal is to simplify the process of creating a table-of-contents (only send heading-structure to the TOC component), and enabling more easy fuzzy-search implementation for markdown headings (in my case, utilizing Fuse.js), and lastly can navigate directly at any heading level locally even without internet connection.

For now, I can only provide ideas, so feel free to modify this feature request and submit a PR (if this proposal is approved) or even enhance this feature further!

## Demo Examples

- Add "headings" array into content-collection result for each markdown
- Demo TOC from the "headings" array
- Demo Fuzzy Search from the "headings" array

To trigger an error, modify the code by changing h1 to h2, replacing # with ## in the codebase, or altering other nested headings to simulate an incorrect heading structure (removing the parent heading). Observe the error in the terminal and within Astro's Error Handler. I believe the main benefit when we can search deeply within markdown headings (e.g., jump to various content headings using the global search component).

DEMO: [StackBlitz](https://stackblitz.com/edit/github-fkmcum)

**Comment by florian-lefebvre** [comment]
I think that's interesting! However, it would be great to have it as a warning (or at least be able to make it a warning in the Astro config). I'm thinking about huge migrations from other tools like wordpress, and having everything break at first doesn't seem a good dx to me.
